pkg/storegateway/indexheader/stream_binary_reader_test.go

```diff
@@ -93,3 +95,32 @@ func TestStreamBinaryReader_CheckSparseHeadersCorrectnessExtensive(t *testing.T)
 		}
 	}
 }
+
+func TestStreamBinaryReader_LabelValuesOffsetsHonorsContextCancel(t *testing.T) {
+	ctx := context.Background()
+
+	tmpDir := filepath.Join(t.TempDir(), "test-sparse index headers")
+	bkt, err := filesystem.NewBucket(filepath.Join(tmpDir, "bkt"))
+	require.NoError(t, err)
+	t.Cleanup(func() { require.NoError(t, bkt.Close()) })
+
+	seriesCount := streamindex.CheckContextEveryNIterations * 10
+	// Create block.
+	lbls := make([]labels.Labels, 0, seriesCount)
+	for i := 0; i < seriesCount; i++ {
+		lbls = append(lbls, labels.FromStrings("a", fmt.Sprintf("%d", i)))
+	}
+	blockID, err := block.CreateBlock(ctx, tmpDir, lbls, 1, 0, 10, labels.FromStrings("ext1", "1"))
+	require.NoError(t, err)
+	require.NoError(t, block.Upload(ctx, log.NewNopLogger(), bkt, filepath.Join(tmpDir, blockID.String()), nil))
+
+	// Write sparse index headers to disk on first build.
+	r, err := NewStreamBinaryReader(ctx, log.NewNopLogger(), bkt, tmpDir, blockID, 3, NewStreamBinaryReaderMetrics(nil), Config{})
+	require.NoError(t, err)
+
+	failAfter := uint64(seriesCount / 2 / streamindex.CheckContextEveryNIterations)
```

Comment:
What's the significance of this number, ensuring the context fails before the first check? Can you add a comment explaining it.

Reply:
added comment